File: algorithm_leetcode/src/main/java/com/huldar/java/day08/LetterCombinations.java
package com.huldar.java.day08;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

/**
 * 电话号码组合
 *
 * @author huldar
 * @date 2019/7/23 16:17
 */
public class LetterCombinations {

    private HashMap<String, String> mapping = new HashMap<String, String>() {{
        put("2", "abc");
        put("3", "def");
        put("4", "ghi");
        put("5", "jkl");
        put("6", "mno");
        put("7", "pqrs");
        put("8", "tuv");
        put("9", "wxyz");
    }};

    private List<String> output = new ArrayList<>();

    /**
     * 使用论坛做法
     *
     * @param digits
     * @return
     */
    public static List<String> letterCombinations1(String digits) {
        LinkedList<String> result = new LinkedList<>();
        String reg = "[2-9]+";
        boolean matchesFlag = digits.matches(reg);
        if (!matchesFlag) {
//            throw new IllegalArgumentException("输入的数据应该为2-9的数字");
            return result;
        }

        //数据字典
        String[] mapping = new String[]{"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        result.add("");
        for (int i = 0; i < digits.length(); i++) {
            char ch = digits.charAt(i);
            int x = Character.getNumericValue(ch);
            while (result.peek().length() == i) {
                String t = result.remove();
                for (char s : mapping[x].toCharArray()) {
                    result.add(t + s);
                }
            }
        }
        return result;
    }

    /**
     * 回溯算法 需要查下资料
     *
     * @param combination
     * @param nextDigits
     */
    private void backtrack(String combination, String nextDigits) {
        if (nextDigits.length() == 0) {
            output.add(combination);
        } else {
            String digit = nextDigits.substring(0, 1);
            String letters = mapping.get(digit);
            for (int i = 0; i < letters.length(); i++) {
                String letter = mapping.get(digit).substring(i, i + 1);
                String combination1 = combination + letter;
                String nextDigits1 = nextDigits.substring(1);
                backtrack(combination1, nextDigits1);
            }
        }
    }

    public List<String> letterCombinations(String digits) {
        if (digits.length() != 0) {
            backtrack("", digits);
        }
        return output;
    }

}

File: algorithm_leetcode/src/main/python/com/huldar/python/day08/LetterCombinations.py
#!usr/bin/env python
# -*- coding:utf-8 _*-
"""
@author:huldar
@file: LetterCombinations.py
@time: 2019/07/23
"""


class Solution:
    @staticmethod
    def letter_combinations(digits):
        """
        电话号码组合-回溯算法
        :param digits: 由2-9的数字组成
        :return: 数字对应的字符的组合
        """
        mapping = {'2': ['a', 'b', 'c'],
                   '3': ['d', 'e', 'f'],
                   '4': ['g', 'h', 'i'],
                   '5': ['j', 'k', 'l'],
                   '6': ['m', 'n', 'o'],
                   '7': ['p', 'q', 'r', 's'],
                   '8': ['t', 'u', 'v'],
                   '9': ['w', 'x', 'y', 'z']}
        out = []

        def backtrack(combination, next_digits):
            """
            回溯算法
            """
            if len(next_digits) == 0:
                out.append(combination)
            else:
                for letter in mapping[next_digits[0]]:
                    backtrack(combination + letter, next_digits[1:])

        if digits:
            backtrack('', digits)

    @staticmethod
    def letter(digits):
        """
        巧妙的解法
        :param digits:
        :return:
        """
        if not digits:
            return []
        l1 = [""]
        h = {"2": "abc", "3": "def", "4": "ghi", "5": "jkl", "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"}
        for char in digits:
            l1 = [each + i for each in l1 for i in h[char]]
        return l1
